core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingInterceptor.java

```diff
@@ -129,20 +132,31 @@ private boolean performPartitionCheck(InvocationContext ctx, LocalFlagAffectedCo
    }
 
    private CompletableFuture<Void> handleDataReadCommand(InvocationContext ctx, DataCommand command) {
-      Object key = command.getKey();
       return ctx.onReturn((rCtx, rCommand, rv, throwable) -> {
+         DataCommand dataCommand = (DataCommand) rCommand;
          if (throwable != null) {
-            if (throwable instanceof RpcException && performPartitionCheck(rCtx, ((DataCommand) rCommand))) {
+            if (throwable instanceof RpcException && performPartitionCheck(rCtx, dataCommand)) {
                // We must have received an AvailabilityException from one of the owners.
                // There is no way to verify the cause here, but there isn't any other way to get an invalid
                // get response.
-               throw log.degradedModeKeyUnavailable(key);
+               throw log.degradedModeKeyUnavailable(dataCommand.getKey());
             } else {
+               // If all owners left and we still haven't received the availability update yet,
+               // we get TimeoutException from JGroupsTransport.checkRsps
+               if (throwable instanceof TimeoutException && performPartitionCheck(rCtx, dataCommand)) {
```

Comment:
Shouldn't the `TimeoutException` be eliminated by the previous commit?

If instead this is a real timeout, I don't think we should change the exception - availability exceptions should be immediate, not after a 15s timeout.